### PR TITLE
feat: [P4PU-549] added redis cache

### PR DESCRIPTION
**Pull request by oleksiybozhykntt**
#### List of Changes
 added redis cache
fixed error code response 

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: helm/values.yaml [hunk 1]
```diff
@@ -88,6 +90,8 @@ microservice-chart:
     JWT_TOKEN_PUBLIC_KEY: jwt-public-key
     JWT_TOKEN_ASSISTANCE_ZENDESK_PRIVATE_KEY: jwt-assistance-zendesk-private-key
     WHITE_LIST_USERS: cf-whitelist
+    REDIS_HOST: cache-standalone-host
+    REDIS_PASSWORD: cache-password
```
[the remark this answers is not in the document]
Comment (oleksiybozhykntt) [reply helm/values.yaml:94]: yes

File: src/main/java/it/gov/pagopa/arc/config/RedisConfig.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    Map<String, RedisCacheConfiguration> cacheConfigurations = new HashMap<>();
+    cacheConfigurations.put(CACHE_OAUTH2_STATE,RedisCacheConfiguration.defaultCacheConfig()
+        .entryTtl(Duration.ofSeconds(300))
+        .disableCachingNullValues());
+    cacheConfigurations.put(CACHE_NAME_ACCESS_TOKEN,redisConfiguration(objectMapper,accessTokenExpirationSeconds));
+    return builder -> builder
+        .withInitialCacheConfigurations(cacheConfigurations);
+  }
+
+  private RedisCacheConfiguration redisConfiguration( ObjectMapper objectMapper,int ttl){
```
[the remark this answers is not in the document]
Comment (oleksiybozhykntt) [reply src/main/java/it/gov/pagopa/arc/config/RedisConfig.java:39]: ok

File: src/main/java/it/gov/pagopa/arc/security/InMemoryOAuth2AuthorizationRequestRepository.java [hunk 3]
```diff
@@ -2,23 +2,27 @@
...
 import jakarta.servlet.http.HttpServletResponse;
-import java.util.Map;
-import java.util.concurrent.ConcurrentHashMap;
 import org.springframework.security.oauth2.client.web.AuthorizationRequestRepository;
 import org.springframework.security.oauth2.core.endpoint.OAuth2AuthorizationRequest;
+import org.springframework.stereotype.Service;
 import org.springframework.util.StringUtils;
 
+@Service
 public class InMemoryOAuth2AuthorizationRequestRepository implements
```
[the remark this answers is not in the document]
Comment (oleksiybozhykntt) [reply src/main/java/it/gov/pagopa/arc/security/InMemoryOAuth2AuthorizationRequestRepository.java:11]: ok

File: src/main/java/it/gov/pagopa/arc/security/InMemoryOAuth2AuthorizationRequestRepository.java [hunk 4]
```diff
@@ -38,7 +42,11 @@ public OAuth2AuthorizationRequest removeAuthorizationRequest(HttpServletRequest
     String state = request.getParameter(STATE);
     if (StringUtils.hasText(state)) {
       // Remove and return the authorization request from the map
-      return authorizationRequestMap.remove(state);
+      OAuth2AuthorizationRequest result = loadAuthorizationRequest(request);
+      if (result!=null){
+        oAuth2AuthorizationRequest.delete(state);
+        return result;
+      }
```
[the remark this answers is not in the document]
Comment (oleksiybozhykntt) [reply src/main/java/it/gov/pagopa/arc/security/InMemoryOAuth2AuthorizationRequestRepository.java:49]: the current implementation requires the deleted element

File: src/main/java/it/gov/pagopa/arc/config/RedisConfig.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (antonioT90) [reply src/main/java/it/gov/pagopa/arc/config/RedisConfig.java]: you have set a generic name, but the method is still focused on handling the `IamUserInfoDTO` class, it's not generic!
add the class as a parameter to make this method really generic

File: src/main/java/it/gov/pagopa/arc/security/RedisOAuth2AuthorizationRequestRepository.java [hunk 5]
```diff
@@ -38,7 +45,11 @@ public OAuth2AuthorizationRequest removeAuthorizationRequest(HttpServletRequest
     String state = request.getParameter(STATE);
     if (StringUtils.hasText(state)) {
       // Remove and return the authorization request from the map
-      return authorizationRequestMap.remove(state);
+      OAuth2AuthorizationRequest result = loadAuthorizationRequest(request);
```
Comment (antonioT90) [inline src/main/java/it/gov/pagopa/arc/security/RedisOAuth2AuthorizationRequestRepository.java:48]: why don't you call directly the `oAuth2AuthorizationRequest.get(state)`

File: src/test/java/it/gov/pagopa/arc/security/IdpIntegrationTest.java [hunk 6]
```diff
@@ -94,6 +95,9 @@ class IdpIntegrationTest {
     @Autowired
     private MockMvc mockMvc;
 
+    @Autowired
+    OAuth2StateStoreRepository oAuth2StateStoreRepository;
```
[the remark this answers is not in the document]
Comment (antonioT90) [reply src/test/java/it/gov/pagopa/arc/security/IdpIntegrationTest.java]: Why are you Autowiring it without using it?

File: src/test/java/it/gov/pagopa/arc/security/InMemoryOAuth2AuthorizationRequestRepositoryTest.java [hunk 7]
```diff
@@ -40,9 +47,10 @@ void givenAuthorizationRequestThenSaveIt() {
   }
 
   @Test
-  void givenAuthorizationRequestThenFailCauseStateNotInRequest(){
+  void givenAuthorizationRequestThenFailCauseStateNotInRequest() {
```
[the remark this answers is not in the document]
Comment (antonioT90) [reply src/test/java/it/gov/pagopa/arc/security/InMemoryOAuth2AuthorizationRequestRepositoryTest.java]: [given<Precodinzioni>]When<NomeMetodo>Then<EsitoAtteso, anche semplice OK>

File: src/test/java/it/gov/pagopa/arc/service/TokenStoreServiceImplTest.java [hunk 8]
```diff
@@ -23,39 +22,48 @@ class TokenStoreServiceImplTest {
...
   );
+  private final String token = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ.SflKxwRJSMeKKF2QT4fwpMeJf36POk6yJV_adQssw5c";
+
   @BeforeEach
   void setUp(){
     tokenStoreService = new TokenStoreServiceImpl();
   }
   @Test
   void givenAccessTokenAndUserInfoThenSaveAndRetrieveTheSameData() {
-    String wrongToken = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9";
```
[the remark this answers is not in the document]
Comment (antonioT90) [reply src/test/java/it/gov/pagopa/arc/service/TokenStoreServiceImplTest.java:32]: [given<Precodinzioni>]When<NomeMetodo>Then<EsitoAtteso, anche semplice OK>

File: src/main/java/it/gov/pagopa/arc/security/RedisOAuth2AuthorizationRequestRepository.java (hunk 5 above)
Comment (oleksiybozhykntt) [reply src/main/java/it/gov/pagopa/arc/security/RedisOAuth2AuthorizationRequestRepository.java:48]: ok

File: src/test/java/it/gov/pagopa/arc/service/TokenStoreServiceImplTest.java [hunk 9]
```diff
@@ -1,61 +1,60 @@
...
+import org.junit.jupiter.api.extension.ExtendWith;
 import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
 
+@ExtendWith(MockitoExtension.class)
 class TokenStoreServiceImplTest {
 
   @Mock
   TokenStoreServiceImpl tokenStoreService;
+  private final String token = "token";
```
[the remark this answers is not in the document]
Comment (oleksiybozhykntt) [reply src/test/java/it/gov/pagopa/arc/service/TokenStoreServiceImplTest.java:16]: removed

File: src/main/java/it/gov/pagopa/arc/config/RedisConfig.java [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+        .disableCachingNullValues());
+    cacheConfigurations.put(CACHE_NAME_ACCESS_TOKEN,redisJsonSerializationConfiguration(objectMapper,accessTokenExpirationSeconds,IamUserInfoDTO.class));
+    return builder -> builder
+        .withInitialCacheConfigurations(cacheConfigurations);
+  }
+
+  private RedisCacheConfiguration redisJsonSerializationConfiguration( ObjectMapper objectMapper,int ttl, Class<?> type){
+    return RedisCacheConfiguration.defaultCacheConfig()
+          .serializeValuesWith(RedisSerializationContext.SerializationPair.fromSerializer(new Jackson2JsonRedisSerializer<>(objectMapper,
+              (Class<?>) type)))
```
Comment (antonioT90) [inline src/main/java/it/gov/pagopa/arc/config/RedisConfig.java:42]: why this cast?
```suggestion
              type)))
```

**Review by antonioT90 — CHANGES_REQUESTED** [review]

**Review by antonioT90 — APPROVED** [review]

**Pull request merged**